cli/crates/cli/tests/introspection.rs

```diff
@@ -0,0 +1,2361 @@
+        "data": {
+            "_service": {
+                "sdl": indoc::indoc! {r#"
+                    type Test {
+                      id: ID!
+                    }
+                "#}
+            }
+        }
+    }));
+
+    Mock::given(method("POST"))
+        .and(path("/graphql"))
+        .and(body_json(request))
+        .and(header("x-api-key", "foo"))
+        .respond_with(response)
+        .mount(&server)
+        .await;
+
+    let address = server.address();
+    let url = format!("http://localhost:{}", address.port());
+
+    let output = env.grafbase_introspect(&url, &["x-api-key: foo"]);
+    let stderr = String::from_utf8_lossy(&output.stderr);
+    let stdout = String::from_utf8_lossy(&output.stdout);
+
+    println!("{stderr}");
+    assert!(output.stderr.is_empty());
+
+    insta::assert_snapshot!(&stdout, @r###"
+    Grafbase CLI 0.40.2
+
+    type Test {
+      id: ID!
+    }
+
+    "###);
+}
+
+#[tokio::test(flavor = "multi_thread")]
+async fn two_headers() {
+    let env = Environment::init();
+    let server = MockServer::start().await;
+
+    let request = json!({
+        "query": "query {\n  _service {\n    sdl\n  }\n}    \n",
+        "variables": {}
+    });
+
+    let response = ResponseTemplate::new(200).set_body_json(json!({
+        "data": {
+            "_service": {
+                "sdl": indoc::indoc! {r#"
+                    type Test {
+                      id: ID!
+                    }
+                "#}
+            }
+        }
+    }));
+
+    Mock::given(method("POST"))
+        .and(path("/graphql"))
+        .and(body_json(request))
+        .and(header("x-api-key", "foo"))
+        .and(header("x-other-key", "bar"))
+        .respond_with(response)
+        .mount(&server)
+        .await;
+
+    let address = server.address();
+    let url = format!("http://localhost:{}", address.port());
+
+    let output = env.grafbase_introspect(&url, &["x-api-key: foo", "x-other-key: bar"]);
+    let stderr = String::from_utf8_lossy(&output.stderr);
+    let stdout = String::from_utf8_lossy(&output.stdout);
+
+    println!("{stderr}");
+    assert!(output.stderr.is_empty());
+
+    insta::assert_snapshot!(&stdout, @r###"
+    Grafbase CLI 0.40.2
+
+    type Test {
+      id: ID!
+    }
+
+    "###);
+}
+
+#[tokio::test(flavor = "multi_thread")]
+#[allow(clippy::too_many_lines)]
+async fn standard() {
+    // bless me, Father, for I have sinned...
+
+    let env = Environment::init();
+    let server = MockServer::start().await;
+
+    let request = json!({
+        "query": "query {\n  _service {\n    sdl\n  }\n}    \n",
+        "variables": {}
+    });
+
+    let response = ResponseTemplate::new(200).set_body_json(json!({
+        "errors": ["boijoing"]
+    }));
+
+    Mock::given(method("POST"))
+        .and(path("/graphql"))
+        .and(body_json(request))
+        .respond_with(response)
+        .mount(&server)
+        .await;
+
+    let request = json!({
```

Comment:
Would semi-recommend just starting an actual GraphQL server and introspecting that.

Introspection isn't guaranteed to match this extract string and it'd be a lot less verbose.